### PR TITLE
Implement cosmetic filtering

### PR DESCRIPTION
**Pull request by antonok-edm**
## Submitter Checklist:

Closes https://github.com/brave/brave-browser/issues/5381
Depends on https://github.com/brave/adblock-rust-ffi/issues/3

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

Majority of tests, covering fine-grained API behavior, are implemented in unit tests within https://github.com/brave/adblock-rust.

FFI behavior is validated by tests in https://github.com/brave/adblock-rust-ffi.

End-to-end browser behavior is verified by preloading custom cosmetic rules into the `adblock-rust` engine, loading a test page, and verifying that DOM elements show as expected using `getComputedStyle`.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/brave_extension/extension/brave_extension/content.ts [hunk 1]
```diff
@@ -1,24 +1,93 @@
...
 let target: EventTarget | null
+let genericExceptions: Array<string> | undefined = undefined
+let classIdBuffer: { classes: string[], ids: string[] } = { classes: [], ids: [] }
 
 function getCurrentURL () {
   return window.location.hostname
 }
 
-/*function applyCosmeticFilterMutationObserver (filterList: any) {
+const getClassesAndIds = (function () {
```
Comment (petemill) [inline components/brave_extension/extension/brave_extension/content.ts]: Is there a reason why this is an Immediately Invoked Function Expression that returns another function? I worry that we'll lose out on some v8 optimizations when this is performance sensitive, and I don't see a reason this can't be a straightforward function with the top-level variables global (to this content script).

File: components/brave_extension/extension/brave_extension/background/events/cosmeticFilterEvents.ts [hunk 2]
```diff
@@ -45,6 +46,32 @@ chrome.runtime.onMessage.addListener((msg, sender, sendResponse) => {
...
       break
     }
+    case 'classIdStylesheet': {
+      chrome.braveShields.classIdStylesheet(msg.classes, msg.ids, msg.exceptions, stylesheet => {
+        chrome.tabs.insertCSS({
+          code: stylesheet,
+          cssOrigin: 'user',
+          runAt: 'document_start'
+        })
+      })
```
Comment (petemill) [inline components/brave_extension/extension/brave_extension/background/events/cosmeticFilterEvents.ts]: `)` should be on separate line since `{` is separate line to `(`

File: components/brave_extension/extension/brave_extension/content.ts (hunk 1 above)
Comment (antonok-edm) [reply components/brave_extension/extension/brave_extension/content.ts]: Just for encapsulation - if it works better with global variables that's fine with me

**Comment by antonok-edm** [comment]
> * **`chrome.braveShields.hostnameCosmeticResources`** ⟶ `getCosemeticStylesheetForHostname`
>   _Though perhaps `...StyleRules...` is less confusing since it returns a string and not a javascript StyleSheet object. Also perhaps `create` / `generate` is clearer than `get`? Behind the scenes is it fetching something or doing some heavy generation work?_

This actually returns 3 distinct things dependent on the hostname: the stylesheet in string format, a list of exceptions to generic cosmetic rules, and a script to be injected in the page. I figured just calling it a stylesheet might be confusing. I'd lean towards `...StyleRules...`, I guess. It is querying saved rules from hashmaps according to the domain and assembling them behind the scenes.

By the way, do you know if there is a convenient way to build a Javascript StyleSheet object as part of the API? That sounds like it would make more sense, although it seems like it would be hard to do on the C++ side / without another layer of wrapping functions.

> * **`chrome.braveShields.classIdStylesheet`** ⟶ `getCosmeticStyleSheetForElements`
>   _Same considerations as for `hostnameCosmeticResources`, perhaps `...ForElementSelectors` is clearer as it doesn't accept actual DOM Elements, but that's quite long._

Perhaps `getCosmeticStylesheetForSelectors`? Although, again this is "generated/created" in a similar way.

File: components/brave_extension/extension/brave_extension/background/events/cosmeticFilterEvents.ts (hunk 2 above)
Comment (antonok-edm) [reply components/brave_extension/extension/brave_extension/background/events/cosmeticFilterEvents.ts]: Hmm, not sure what you mean? The first `})` matches to `...insertCSS({`, the second matches to `...Stylesheet(... => {`.

**Comment by antonok-edm** [comment]
Also just added a toggle in settings corresponding to the new shields panel option.

**Comment by szaimen** [comment]
Any update here?

**Comment by antonok-edm** [comment]
@szaimen Still blocked on code reviews for https://github.com/brave/adblock-rust/pull/50 unfortunately

**Comment by szaimen** [comment]
@antonok-edm any update here now that https://github.com/brave/adblock-rust/pull/50 is merged?

**Comment by antonok-edm** [comment]
@szaimen https://github.com/brave/brave-core/pull/3419 and https://github.com/brave/brave-core-crx-packager/pull/74 are required to fully support https://github.com/brave/adblock-rust/pull/50 in the browser. Cosmetic filtering will be unblocked once that's sorted.

File: components/brave_shields/browser/ad_block_base_service.h [hunk 3]
```diff
@@ -45,6 +46,9 @@ class AdBlockBaseService : public BaseBraveShieldsService {
   void EnableTag(const std::string& tag, bool enabled);
   bool TagExists(const std::string& tag);
 
+  adblock::HostnameResources hostnameCosmeticResources(const std::string& hostname);
```
Comment (bbondy) [inline components/brave_shields/browser/ad_block_base_service.h]: I'm a bit confused by this, it looks like the implementation of this function just calls directly into the return type which seems to come from wrapper.hpp. But from what I can tell it should return an `std::string` and not `HostnameResources` here https://github.com/brave/adblock-rust-ffi/pull/7/files#diff-afc8c31c40724bcf2f02d090d196dbedR60

I don't even see `HostnameResources` defined anywhere. I must be missing something, or else there's a missing commit or something.  Please explain ,thanks!

File: components/brave_extension/extension/brave_extension/content.ts [hunk 4]
```diff
@@ -1,10 +1,100 @@
...
+  return { classes, ids }
+}
+
+const handleNewNodes = (newNodes: Element[]) => {
+  const { classes, ids } = getClassesAndIds(newNodes)
+  if (!(genericExceptions)) {
+    classIdBuffer.classes.push(...classes)
+    classIdBuffer.ids.push(...ids)
+  } else {
+    chrome.runtime.sendMessage({
```
Comment (bbondy) [inline components/brave_extension/extension/brave_extension/content.ts]: In practice, do you know how often `classIdStylesheet` gets hit for pages?  Just making sure we're not inserting like 1000 stylesheets.  I think maybe we should be debouncing these calls.

File: components/brave_shields/browser/ad_block_service_browsertest.cc [hunk 5]
```diff
@@ -744,3 +744,112 @@ IN_PROC_BROWSER_TEST_F(AdBlockServiceTest, CancelRequestOptionTest) {
...
+  ASSERT_TRUE(ExecuteScriptAndExtractBool(contents,
+                                          "checkSelector('.ad', 'display', 'block')",
+                                          &as_expected));
+  EXPECT_TRUE(as_expected);
+
+  as_expected = false;
+  ASSERT_TRUE(ExecuteScriptAndExtractBool(contents,
+                                          "checkSelector('#ad-banner', 'display', 'none')",
+                                          &as_expected));
+  EXPECT_TRUE(as_expected);
```
Comment (bbondy) [inline components/brave_shields/browser/ad_block_service_browsertest.cc]: Good job on these tests 👍

File: components/brave_extension/extension/brave_extension/_locales/en_US/messages.json [hunk 6]
```diff
@@ -47,6 +47,10 @@
     "message": "connection upgraded",
     "description": "Message for the main blocked resources text when there is at least one ad/track/script blocked and one connection upgrade (singular)"
   },
+  "firstPartyAdsBlocked": {
```
Comment (bbondy) [inline components/brave_extension/extension/brave_extension/_locales/en_US/messages.json]: pls change this ID name to be cosmetic instead of firstParty

File: components/brave_shields/browser/ad_block_service_browsertest.cc [hunk 7]
```diff
@@ -744,3 +744,112 @@ IN_PROC_BROWSER_TEST_F(AdBlockServiceTest, CancelRequestOptionTest) {
...
+  as_expected = false;
+  ASSERT_TRUE(ExecuteScriptAndExtractBool(contents,
+                                          "checkSelector('.ad', 'display', 'none')",
+                                          &as_expected));
+  EXPECT_TRUE(as_expected);
+}
+
+// Test cosmetic filtering on elements added dynamically
+IN_PROC_BROWSER_TEST_F(AdBlockServiceTest, CosmeticFilteringDynamic) {
+  UpdateAdBlockInstanceWithRules(
```
Comment (bbondy) [inline components/brave_shields/browser/ad_block_service_browsertest.cc]: You might want a call to WaitForBraveExtensionAdded at the start above here.

File: browser/resources/settings/default_brave_shields_page/default_brave_shields_page.html [hunk 8]
```diff
@@ -90,6 +90,11 @@
         </settings-radio-group>
       </div>
     </div>
+    <settings-toggle-button id="cosmeticFilteringControlType"
+        pref="{{prefs.brave.cosmetic_filtering_default}}"
+        label="$i18n{cosmeticFilteringControlLabel}"
```
Comment (bbondy) [inline browser/resources/settings/default_brave_shields_page/default_brave_shields_page.html]: Could you put a screenshot for how this looks in the PR? Thanks!

File: components/brave_shields/browser/ad_block_service_browsertest.cc (hunk 5 above)
Comment (bbondy) [reply components/brave_shields/browser/ad_block_service_browsertest.cc]: Could you add a test for when cosmetic filters are off that it doesn't actually do any work?
Either by preference or by calling this explicitly:
```
brave_shields::SetCosmeticFilteringControlType(profile_,
                                  value ? ControlType::BLOCK
                                        : ControlType::ALLOW,
                                  GURL());
```
I think setting just the pref won't work currently because this is done in the callback and the pref isn't actually checked anywhere.

File: components/brave_extension/extension/brave_extension/background/api/cosmeticFilterAPI.ts [hunk 9]
```diff
@@ -22,7 +22,27 @@ export const removeSiteFilter = (origin: string) => {
...
+      cssOrigin: 'user',
+      runAt: 'document_start'
+    })
+
+    chrome.tabs.sendMessage(tabId, {
+      type: 'cosmeticFilterGenericExceptions',
+      exceptions: genericExceptions
+    })
+
+    chrome.tabs.executeScript(tabId, {
```
[the remark this answers is not in the document]
Comment (antonok-edm) [reply components/brave_extension/extension/brave_extension/background/api/cosmeticFilterAPI.ts]: The scripts come from the same set of resources as for network redirects.
There are a few extra parameterized templated scriptlets, which were included in the same security review.

Here's an example of a scriptlet injection rule with 1 argument, directly from uBlock's filters:
```
computerbild.de##+js(abort-on-property-read.js, Date.prototype.toUTCString)
```

The argument (`Date.prototype.toUTCString`) is interpolated into [this function](https://github.com/gorhill/uBlock/blob/master/assets/resources/scriptlets.js#L96) at the location of the first occurrence of `{{1}}`, and the resulting function would be used in this `executeScript` call.

File: components/brave_extension/extension/brave_extension/background/reducers/shieldsPanelReducer.ts [hunk 10]
```diff
@@ -331,6 +361,19 @@ export default function shieldsPanelReducer (
       })
       break
     }
+    case shieldsPanelTypes.PAGE_CONTENT_READY_FOR_INJECTION: {
+      const tabData = state.tabs[action.tabId]
+      let cosmeticBlockingEnabled: boolean
+      if (!tabData) {
+        console.warn('Active tab not found')
+        cosmeticBlockingEnabled = true
```
[the remark this answers is not in the document]
Comment (antonok-edm) [reply components/brave_extension/extension/brave_extension/background/reducers/shieldsPanelReducer.ts]: I ran into it during automated testing and couldn't figure out how else to run test cases with cosmetic filtering enabled. I've never seen it during normal browsing though. Would appreciate help with a more "correct" solution.

File: components/brave_shields/browser/ad_block_base_service.h (hunk 3 above)
Comment (antonok-edm) [reply components/brave_shields/browser/ad_block_base_service.h]: Sorry, that didn't get updated since changing the FFI interface. You are correct in that it should be a `std::string`.

File: components/brave_shields/browser/ad_block_service_browsertest.cc (hunk 5 above)
Comment (antonok-edm) [reply components/brave_shields/browser/ad_block_service_browsertest.cc]: For the reason I mentioned [above](https://github.com/brave/brave-core/pull/3303#discussion_r348112733), I couldn't actually get the value of the preference when running tests, so I defaulted to having cosmetic filtering on. This would be a good thing to fix and test, but I don't know where to go from here.

File: components/brave_shields/browser/ad_block_service_browsertest.cc (hunk 7 above)
Comment (antonok-edm) [reply components/brave_shields/browser/ad_block_service_browsertest.cc]: Hmm, that may help with https://github.com/brave/brave-core/pull/3303#discussion_r348114676... will try

File: browser/resources/settings/default_brave_shields_page/default_brave_shields_page.html (hunk 8 above)
Comment (tomlowenthal) [reply browser/resources/settings/default_brave_shields_page/default_brave_shields_page.html]: Will need to take a look at the design and especially strings for this.

File: components/brave_extension/extension/brave_extension/content.ts [hunk 11]
```diff
@@ -1,10 +1,100 @@
...
+
+const queriedIds = new Set()
+const queriedClasses = new Set()
+const regexWhitespace = /\s/
 
 function getCurrentURL () {
   return window.location.hostname
 }
 
+const getClassesAndIds = function (addedNodes: Element[]) {
```
[the remark this answers is not in the document]
Comment (antonok-edm) [reply components/brave_extension/extension/brave_extension/content.ts]: This is what uBlock Origin does and I haven't found the performance hit to be noticeable in practice, but it would be nice to take advantage of native APIs if possible.

**Comment by tomlowenthal** [comment]
I see that there's code for a new toggle in the Shields panel. Is a corresponding control added in the shields area of settings?

**Comment by antonok-edm** [comment]
@tomlowenthal Yes, and I'll post screenshots of both once I get a build up

File: browser/resources/settings/default_brave_shields_page/default_brave_shields_page.html (hunk 8 above)
Comment (antonok-edm) [reply browser/resources/settings/default_brave_shields_page/default_brave_shields_page.html]: FWIW, I based the implementation from [this mockup](https://app.goabstract.com/projects/bad752b0-846f-11e8-aef1-25c293be1275/branches/d7aacab1-fa2f-4adf-a0db-3b0eb43098ec/commits/413fc3f9a676ccca135f229fc3aac16a84b9e258/files/47C7B326-36ED-428B-A70C-006F108E6A50/layers/D93DE291-D528-4CC4-B6A7-115DFF48E61E?mode=design)

File: components/brave_extension/extension/brave_extension/content.ts (hunk 4 above)
Comment (antonok-edm) [reply components/brave_extension/extension/brave_extension/content.ts]: It depends entirely on how often the site makes mutations to the page. I can try on a few different ones if you'd like specific examples.
Note that this is already batched on a per-mutation basis - i.e. if a node tree is added, it's only called once for all the classes and ids in the entire tree.

**Comment by antonok-edm** [comment]
@bbondy @tomlowenthal

Shields panel screenshot:
2019-11-19-133919_542x852_scrot

Settings screenshot:
2019-11-19-134135_718x706_scrot

File: browser/resources/settings/default_brave_shields_page/default_brave_shields_page.html (hunk 8 above)
Comment (antonok-edm) [reply browser/resources/settings/default_brave_shields_page/default_brave_shields_page.html]: Posted in a separate comment below.

File: components/brave_shields/browser/ad_block_service_browsertest.cc (hunk 7 above)
Comment (antonok-edm) [reply components/brave_shields/browser/ad_block_service_browsertest.cc]: That doesn't seem to help, unfortunately.

Was there another reason you recommended adding that? It wasn't implemented on any of the other tests in the file either.

File: components/brave_shields/browser/ad_block_service_browsertest.cc (hunk 5 above)
Comment (antonok-edm) [reply components/brave_shields/browser/ad_block_service_browsertest.cc]: Since moving to a feature flag I was able to make a test for this.

File: patches/chrome-browser-flag_descriptions.cc.patch [hunk 12]
```diff
@@ -0,0 +1,15 @@
+diff --git a/chrome/browser/flag_descriptions.cc b/chrome/browser/flag_descriptions.cc
+index 62fa3e5cb73b777cef2e0272fe2d892f352f18df..b083f1ec5f5691d147705dc15ba27f09e2e4d8d3 100644
+--- a/chrome/browser/flag_descriptions.cc
++++ b/chrome/browser/flag_descriptions.cc
+@@ -2000,6 +2000,10 @@ const char kSystemKeyboardLockDescription[] =
+     "keyboard shortcuts and have the events routed directly to the website "
+     "when in fullscreen mode.";
+ 
++const char kBraveAdblockCosmeticFilteringName[] = "Enable cosmetic filtering";
```
Comment (bridiver) [inline patches/chrome-browser-flag_descriptions.cc.patch]: this and the header file below can be done with chromium_src overrides. See `chromium_src/content/public/common/url_constants.cc`

File: patches/chrome-browser-about_flags.cc.patch [hunk 13]
```diff
@@ -1,8 +1,17 @@
...
-index 1518449679e8c4a457a58042b5a0f6a70f385db9..15631077869506e67b36a4d25408917cd9cb92d8 100644
+index 1518449679e8c4a457a58042b5a0f6a70f385db9..1ece0b4cb755e02937c478df918fd31df947052c 100644
 --- a/chrome/browser/about_flags.cc
 +++ b/chrome/browser/about_flags.cc
-@@ -282,14 +282,14 @@ const FeatureEntry::Choice kTouchTextSelectionStrategyChoices[] = {
+@@ -164,6 +164,8 @@
+ #include "ui/gl/gl_switches.h"
+ #include "ui/native_theme/native_theme_features.h"
+ 
++#include "brave/browser/features.h"
```
Comment (bridiver) [inline patches/chrome-browser-about_flags.cc.patch]: I would use a chromium_src override to add this to chrome_features.h

File: patches/chrome-browser-about_flags.cc.patch [hunk 14]
```diff
@@ -1,8 +1,17 @@
...
++
```
Comment (bridiver) [inline patches/chrome-browser-about_flags.cc.patch]: never add blank lines in a patch

File: components/brave_extension/extension/brave_extension/background/api/cosmeticFilterAPI.ts [hunk 15]
```diff
@@ -22,7 +38,37 @@ export const removeSiteFilter = (origin: string) => {
   })
 }
 
-export const applySiteFilters = (tabId: number, hostname: string) => {
+export const applyAdblockCosmeticFilters = (tabId: number, hostname: string) => {
```
Comment (bridiver) [inline components/brave_extension/extension/brave_extension/background/api/cosmeticFilterAPI.ts]: this is really not the right way to do this and the same thing applies here https://github.com/brave/brave-core/pull/3303/files#diff-3493bfe26c7890d60718508a5969519aR84
we should be using the declarative content api to add and remove rules when the settings change https://developer.chrome.com/extensions/declarativeContent

File: components/brave_shields/browser/ad_block_base_service.h [hunk 16]
```diff
@@ -46,6 +48,13 @@ class AdBlockBaseService : public BaseBraveShieldsService {
   void EnableTag(const std::string& tag, bool enabled);
   bool TagExists(const std::string& tag);
 
+  base::Optional<base::Value> hostnameCosmeticResources(
```
Comment (bridiver) [inline components/brave_shields/browser/ad_block_base_service.h]: lint nit - method names should start with capital letters

File: components/brave_shields/browser/ad_block_base_service.h [hunk 17]
```diff
@@ -13,12 +13,14 @@
...
 #include <vector>
 
+#include "base/values.h"
 #include "base/files/file_path.h"
 #include "base/memory/weak_ptr.h"
 #include "base/sequence_checker.h"
 #include "brave/components/brave_shields/browser/base_brave_shields_service.h"
 #include "brave/components/brave_component_updater/browser/dat_file_util.h"
 #include "content/public/common/resource_type.h"
+#include "brave/vendor/adblock_rust_ffi/src/wrapper.hpp"
```
Comment (bridiver) [inline components/brave_shields/browser/ad_block_base_service.h]: this will fail lint for alphabetical order

Comment (bridiver) [reply components/brave_shields/browser/ad_block_base_service.h]: and actually it should only be in the `.cc` file and not the header

Comment (bridiver) [reply components/brave_shields/browser/ad_block_base_service.h]: actually I'm not seeing why this is needed at all

File: browser/extensions/api/brave_shields_api.cc [hunk 18]
```diff
@@ -23,8 +23,12 @@
 #include "chrome/browser/extensions/chrome_extension_function_details.h"
 #include "chrome/browser/extensions/extension_tab_util.h"
 #include "chrome/browser/profiles/profile.h"
+#include "chrome/common/chrome_features.h"
 #include "content/public/browser/web_contents.h"
 #include "extensions/browser/extension_util.h"
+#include "brave/components/brave_shields/browser/ad_block_base_service.h"
```
Comment (bridiver) [inline browser/extensions/api/brave_shields_api.cc]: why is this needed?

File: components/brave_shields/browser/ad_block_base_service.h [hunk 19]
```diff
@@ -46,6 +48,13 @@ class AdBlockBaseService : public BaseBraveShieldsService {
   void EnableTag(const std::string& tag, bool enabled);
   bool TagExists(const std::string& tag);
 
+  base::Optional<base::Value> hostnameCosmeticResources(
+          const std::string& hostname);
+  std::string classIdStylesheet(
```
Comment (bridiver) [inline components/brave_shields/browser/ad_block_base_service.h]: same as above

File: browser/extensions/api/brave_shields_api.cc [hunk 20]
```diff
@@ -23,8 +23,12 @@
...
+#include "brave/components/brave_shields/browser/ad_block_service.h"
```
Comment (bridiver) [inline browser/extensions/api/brave_shields_api.cc]: you need a dep for`//brave/components/brave_shields/browser` in `brave/browser/extensions/BUILD.gn`

File: browser/extensions/api/brave_shields_api.cc [hunk 21]
```diff
@@ -23,8 +23,12 @@
 #include "chrome/browser/extensions/chrome_extension_function_details.h"
 #include "chrome/browser/extensions/extension_tab_util.h"
 #include "chrome/browser/profiles/profile.h"
+#include "chrome/common/chrome_features.h"
```
Comment (bridiver) [inline browser/extensions/api/brave_shields_api.cc]: you need a dep for `//chrome/common` in brave/browser/extensions/BUILD.gn

File: components/brave_extension/extension/brave_extension/background/reducers/shieldsPanelReducer.ts [hunk 22]
```diff
@@ -340,6 +347,45 @@ export default function shieldsPanelReducer (
...
     }
+    case shieldsPanelTypes.GENERATE_CLASS_ID_STYLESHEET: {
+      const tabData = state.tabs[action.tabId]
+      if (!tabData) {
+        console.error('Active tab not found')
+        break
+      }
+      const exceptions = tabData.cosmeticFilters.ruleExceptions
+
+      injectClassIdStylesheet(action.tabId, action.classes, action.ids, exceptions)
```
Comment (petemill) [inline components/brave_extension/extension/brave_extension/background/reducers/shieldsPanelReducer.ts]: It could be a good idea to wrap this side-effect in a `setTimeout(fn, 0)` since that function could call another redux function immediately. I know it's inside a callback, but for all we know that callback function could get called immediately, before the current redux action is completed.

**Comment by petemill** [comment]
As an aside, is there anywhere documented as to what the various cosmetic filter properties are? In other words, what 'exceptions' are, as opposed to either selectors for hiding, custom CSS rules, custom JS script and whatever else is contained.

File: DEPS [hunk 23]
```diff
@@ -1,7 +1,7 @@
 use_relative_paths = True
 
 deps = {
-  "vendor/adblock_rust_ffi": "https://github.com/brave/adblock-rust-ffi.git@89127a30655eaf54cf73794309846084ea8b91b9",
+  "vendor/adblock_rust_ffi": "https://github.com/brave/adblock-rust-ffi.git@6dad8c1b47c1b866fa1b1c19e36da6eb5797a524",
```
Comment (bbondy) [inline DEPS]: Before this gets merged please merge this: 
https://github.com/brave/adblock-rust-ffi/pull/7

And then update this changeset ref. As I only want to commit the ref here for things in master on the ffi repo.

File: browser/extensions/api/brave_shields_api.cc (hunk 21 above)
Comment (bbondy) [reply browser/extensions/api/brave_shields_api.cc]: I verified it's there now. Resolving this.

**Comment by antonok-edm** [comment]
@petemill 

> As an aside, is there anywhere documented as to what the various cosmetic filter properties are? In other words, what 'exceptions' are, as opposed to either selectors for hiding, custom CSS rules, custom JS script and whatever else is contained.

Most of how this works is explained in doc comments from `adblock-rust`. I just double checked and there are a few important data structures that could use some additional explanation, so I'll follow up there as well.

**Comment by bridiver** [comment]
Jenkins is failing because of some glibc error that does not appear to be related to this PR

**Comment by bridiver** [comment]
I'm approving this as ok to merge because @antonok-edm has verified the build on linux

**Comment by bsclifton** [comment]
Reverted with https://github.com/brave/brave-core/pull/4255

Note: when this was merged, milestone should have been set @antonok-edm (ex: set milestone to 1.4, set milestone for issue also). I re-opened the issue and left it without a milestone
